openwisp_controller/config/base/device.py

```diff
@@ -111,6 +111,10 @@ def clean(self):
         if self._state.adding:
             return
         current = self.__class__.objects.get(pk=self.pk)
+        if self.name != current.name:
+            device_name_changed.send(
+                sender=self.__class__, instance=self,
+            )
```

Comment:
I wonder why this logic was added to the clean method instead of `save()`. Give me some time to clear this doubt. I think it may be better to move these checks (including the one which comes afterwards) in the save method.

Reply:
Ok Should I wait or shift the `clean()` method code into `save()` method ?

Even I think `save()` will be a better option for this.

cc: @nemesisdesign

Reply:
After reading ["Validating" objects section of Model instance reference in Django Documentation](https://docs.djangoproject.com/en/3.1/ref/models/instances/#validating-objects), but gut says to call `full_clean` in `save` method. 

On second thought, doing so will call `full_clean` twice when a user saves a Device object from Django admin. Once by `ModelForm` and other by `Device.save` method.

Reply:
I have fixed this issue in a separate patch to keep things saner: https://github.com/openwisp/openwisp-controller/pull/409
Please rebase on the current master and resolve conflicts, thanks!